runtime/contract_update_validation.go

```diff
@@ -0,0 +1,434 @@
+	// only leave some unused-data, and will not do any harm to the programs that are running.
+
+	// This is a fail-fast check.
+	if len(oldFields) < len(newFields) {
+		if validator.isInNestedDecl() {
+			return fmt.Errorf("too many fields in `%s`. expected %d, found %d",
+				validator.currentDeclName, len(oldFields), len(newFields))
+		}
+
+		return fmt.Errorf("too many fields. expected %d, found %d",
+			len(oldFields), len(newFields))
+	}
+
+	// Put the old field-types against their field-names to a map, for faster lookup.
+	oldFiledTypes := map[string]ast.Type{}
+	for _, field := range oldFields {
+		oldFiledTypes[field.Identifier.Identifier] = field.TypeAnnotation.Type
+	}
+
+	for _, newField := range newFields {
+		err := validator.visitField(newField, oldFiledTypes)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (validator *ContractUpdateValidator) visitField(
+	newField *ast.FieldDeclaration,
+	oldFiledTypes map[string]ast.Type) error {
+
+	prevFieldName := validator.currentFieldName
+	validator.currentFieldName = newField.Identifier.Identifier
+	defer func() {
+		validator.currentFieldName = prevFieldName
+	}()
+
+	oldType := oldFiledTypes[validator.currentFieldName]
+	if oldType == nil {
+		if validator.isInNestedDecl() {
+			return fmt.Errorf("found new field `%s` in `%s",
+				validator.currentDeclName,
+				validator.currentFieldName)
+		}
+
+		return fmt.Errorf("found new field `%s`", validator.currentFieldName)
+	}
+
+	return oldType.Equal(newField.TypeAnnotation.Type, validator)
+}
+
+func (validator *ContractUpdateValidator) CheckNominalTypeEquality(expected *ast.NominalType, found ast.Type) error {
+	foundNominalType, ok := found.(*ast.NominalType)
+	if !ok {
+		return validator.getTypeMismatchError(expected, found)
+	}
+
+	// First check whether the names are equal.
+	ok = validator.checkNameEquality(expected, foundNominalType)
+	if !ok {
+		return validator.getTypeMismatchError(expected, found)
+	}
+
+	// Then, check whether the fields of the two composite declarations
+	// referred by this nominal type, are also compatible.
+
+	var compositeDeclName string
+	if isQualifiedName(foundNominalType) {
+		compositeDeclName = foundNominalType.NestedIdentifiers[0].Identifier
+	} else {
+		compositeDeclName = foundNominalType.Identifier.Identifier
+	}
+
+	validator.loadCompositeDecls()
+
+	oldCompositeDecl := validator.oldNestedCompositeDecls[compositeDeclName]
+	if oldCompositeDecl == nil {
+		// If the declaration is not available, that means this is an imported contract.
+		// Thus, no need to validate anymore.
+		return nil
+	}
+
+	newCompositeDecl := validator.newNestedCompositeDecls[compositeDeclName]
+	if newCompositeDecl == nil {
+		// If the declaration is not available, that means this is an imported contract.
+		// Thus, no need to validate anymore. Ideally shouldn't reach here.
+		return nil
+	}
+
+	return validator.checkDeclarationUpdatability(oldCompositeDecl, newCompositeDecl)
```

Comment:
I don't think this check is needed here. `CheckNominalTypeEquality` should just check the type equality.

The checking of nested composites declarations should recurse down into checking nested declaration.

Reply:
Yeah, I think we should recursively visit nested declarations as well. That will solve [this problem](https://github.com/onflow/cadence/issues/588#issuecomment-776628175) as well.